mlflow/models/evaluation/base.py

```diff
@@ -0,0 +1,559 @@
+        )
+        md5_gen.update(len_bytes)
+        if len(data) < EvaluationDataset.NUM_SAMPLE_ROWS_FOR_HASH * 2:
+            md5_gen.update(EvaluationDataset._array_like_obj_to_bytes(data))
+        else:
+            md5_gen.update(
+                EvaluationDataset._array_like_obj_to_bytes(
+                    data[: EvaluationDataset.NUM_SAMPLE_ROWS_FOR_HASH]
+                )
+            )
+            md5_gen.update(
+                EvaluationDataset._array_like_obj_to_bytes(
+                    data[-EvaluationDataset.NUM_SAMPLE_ROWS_FOR_HASH :]
+                )
+            )
+
+    @property
+    def name(self):
+        """
+        Dataset name, which is specified dataset name or the dataset hash if user don't specify
+        name.
+        """
+        return self._user_specified_name if self._user_specified_name is not None else self.hash
+
+    @property
+    def hash(self):
+        """
+        Compute a hash from the specified dataset by selecting the first 5 records, last 5 records,
+        dataset size and feeding them through a cheap, low-collision hash function
+        """
+        import numpy as np
+        import pandas as pd
+
+        if self._hash is None:
+            md5_gen = hashlib.md5()
+            if isinstance(self.data, np.ndarray):
+                EvaluationDataset._gen_md5_for_arraylike_obj(md5_gen, self.data)
+                EvaluationDataset._gen_md5_for_arraylike_obj(md5_gen, self.labels)
+            elif isinstance(self.data, pd.DataFrame):
+                column_names = ",".join(self.data.columns)
+                meta_str = f"columns={column_names}\nlabels={self.labels}"
+                md5_gen.update(meta_str.encode("UTF-8"))
+                EvaluationDataset._gen_md5_for_arraylike_obj(md5_gen, self.data)
+            self._hash = md5_gen.hexdigest()
+        return self._hash
+
+    @property
+    def _metadata(self):
+        """
+        Return dataset metadata containing name, hash, and optional path.
+        """
+        metadata = {
+            "name": self.name,
+            "hash": self.hash,
+        }
+        if self.path is not None:
+            metadata["path"] = self.path
+        return metadata
+
+    def _log_dataset_tag(self, client, run_id):
+        """
+        Log dataset metadata as a tag "mlflow.datasets", if the tag already exists, it will
+        append current dataset metadata into existing tag content.
+        """
+        existing_dataset_metadata_str = client.get_run(run_id).data.tags.get(
+            "mlflow.datasets", "[]"
+        )
+        dataset_metadata_list = json.loads(existing_dataset_metadata_str)
+
+        for metadata in dataset_metadata_list:
+            if metadata["hash"] == self.hash and metadata["name"] == self._user_specified_name:
+                break
+        else:
+            dataset_metadata_list.append(self._metadata)
+
+        dataset_metadata_str = json.dumps(dataset_metadata_list)
```

Comment:
Can we remove spaces from the datasets JSON so that it's easier to write MLflow Search queries with ILIKE syntax? This way, we don't have to write queries that assume there's a single space between each attribute type (e.g. `name`) and the attribute value (e.g. `my_dataset`).

```suggestion
        dataset_metadata_str = json.dumps(dataset_metadata_list, separators=(',', ':'))
```

Should do it - reference https://stackoverflow.com/a/16311587/11952869.

Can we also add a test case to make sure there isn't any whitespace in the `mlflow.datasets` tag content?